### DISCUSSION TITLE
Inheriting environment variables from dependency tasks

### DISCUSSION THREAD

**Discussion by postlund** [go-task/task#705] [Q&A]
So, I have this use case: I want to build firmware for a particular SoC using a cross-compiler and support doing so on any operating system (as our developers currently run Windows and macOS but we also use Linux for CI). I have pre-built binaries of the compilers for each platforms and pre-existing makefiles for building, so what I basically need is a way to tie everything together. There's a lot of backend, frontend, test code and other tools that will move into this system later as well.

My intended way to implement this is to create OS-specific tasks to download and unpack the compiler needed, depending on what operating system Taskfile is run on. The task would update PATH to include binaries in the unpacked files. Finally I would just want to depend on this task for each firmware I build (since I have multiple) and just re-use the environment (to run `make`). As far as I understand, it's not possible to carry over (environment) variables from a task to its dependents? Any other suggestions for doing this?

(Platform independent support to download external files in a Taskfile would be great btw)

**Comment by ghostsquad** [comment]
For clarity, when using an acronym, please spell it out once, the first time you use it. e.g. 

> SoC (Security Operation Center)
> SoC (System on a Chip)

It's probably ok to use some very very common acronyms like `OS`, even `CI/CD`, but otherwise, it may be hard for folks to follow along. Thank you for understanding!

---

> it's not possible to carry over (environment) variables from a task to its dependents

I believe that you must _explicitly_ pass `env` vars (as well as "regular" `var` vars) from parent task to deps (or tasks in `cmds`) as per the documentation here: https://taskfile.dev/#/usage?id=task-dependencies but @andreynering would know best.

I'm tagging this as `v4` cause I believe the use case here matches quite well with the `matrix` feature I'm working on (described more here: https://github.com/go-task/task/discussions/700#discussioncomment-2527969)

---

> (Platform independent support to download external files in a Taskfile would be great btw)

A `task` package manager (more or less) is also on my radar for v4. Technically, I won't be specifically writing any code that does platform specific lookup/work, however, like with [asdf plugins](https://github.com/particledecay/asdf-task/blob/master/bin/download#L27) there will be built-in vars & functions you can use to dynamically create the URL for which external Taskfile(s) you wish to download.

**Reply by postlund** [reply]
> For clarity, when using an acronym, please spell it out once, the first time you use it. e.g.

Absolutely! 👍 

> I believe that you must _explicitly_ pass `env` vars (as well as "regular" `var` vars) from parent task to deps (or tasks in `cmds`) as per the documentation here: https://taskfile.dev/#/usage?id=task-dependencies but @andreynering would know best.

I believe so too, it's a bit inconvenient to use though.

> I'm tagging this as `v4` cause I believe the use case here matches quite well with the `matrix` feature I'm working on (described more here: [#700 (comment)](https://github.com/go-task/task/discussions/700#discussioncomment-2527969))

That actually sounds like a great addition!

> A `task` package manager (more or less) is also on my radar for v4. Technically, I won't be specifically writing any code that does platform specific lookup/work, however, like with [asdf plugins](https://github.com/particledecay/asdf-task/blob/master/bin/download#L27) there will be built-in vars & functions you can use to dynamically create the URL for which external Taskfile(s) you wish to download.

In my case it's not the URLs that are the problem, it's downloading the content since I don't know what tools are actually available on the machine running `task.`. I can bundle tools with in my repo, sure, but I don't feel like that is the optimal solution.

**Reply by ghostsquad** [reply]
What I meant is that, I plan on including a way to specify a url to a Taskfile you wish to download (and allow the use of templating to make it dynamic), and then downloading the file for you (along with checksum verification functionality). So you won't need to know if `curl` is installed or anything like that.
